File: epi_judge_java_solutions/epi/IsAnonymousLetterConstructible.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class IsAnonymousLetterConstructible {
  @EpiTest(testDataFile = "is_anonymous_letter_constructible.tsv")

  public static boolean isLetterConstructibleFromMagazine(String letterText,
                                                          String magazineText) {

    // Compute the frequencies for all chars in letterText.
    Map<Character, Long> charFrequencyForLetter =
        letterText.chars()
            .mapToObj(c -> (char)c)
            .collect(Collectors.groupingBy(Function.identity(),
                                           Collectors.counting()));

    // Check if the characters in magazineText can cover characters in
    // letterText.
    for (char c : magazineText.toCharArray()) {
      if (charFrequencyForLetter.containsKey(c)) {
        charFrequencyForLetter.put(c, charFrequencyForLetter.get(c) - 1);
        if (charFrequencyForLetter.remove(c, 0L)) {
          charFrequencyForLetter.remove(c);
          // All characters for letterText are matched.
          if (charFrequencyForLetter.isEmpty()) {
            break;
          }
        }
      }
    }
    // Empty charFrequencyForLetter means every char in letterText can be
    // covered by a character in magazineText.
    return charFrequencyForLetter.isEmpty();
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "IsAnonymousLetterConstructible.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_cpp_solutions/is_anonymous_letter_constructible.cc
#include <string>
#include <unordered_map>

#include "test_framework/generic_test.h"

using std::string;
using std::unordered_map;

bool IsLetterConstructibleFromMagazine(const string& letter_text,
                                       const string& magazine_text) {
  unordered_map<char, int> char_frequency_for_letter;
  // Compute the frequencies for all chars in letter_text.
  for (char c : letter_text) {
    ++char_frequency_for_letter[c];
  }

  // Check if the characters in magazine_text can cover characters
  // in char_frequency_for_letter.
  for (char c : magazine_text) {
    if (auto it = char_frequency_for_letter.find(c);
        it != cend(char_frequency_for_letter)) {
      --it->second;
      if (it->second == 0) {
        char_frequency_for_letter.erase(it);
        if (empty(char_frequency_for_letter)) {
          // All characters for letter_text are matched.
          break;
        }
      }
    }
  }
  // Empty char_frequency_for_letter means every char in letter_text can be
  // covered by a character in magazine_text.
  return empty(char_frequency_for_letter);
}

// clang-format off


int main(int argc, char* argv[]) {
  std::vector<std::string> args {argv + 1, argv + argc};
  std::vector<std::string> param_names {"letter_text", "magazine_text"};
  return GenericTestMain(args, "is_anonymous_letter_constructible.cc", "is_anonymous_letter_constructible.tsv", &IsLetterConstructibleFromMagazine,
                         DefaultComparator{}, param_names);
}
// clang-format on
